### PR TITLE
Np 581 read crossref data

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: config/pmd/ruleset.xml [hunk 1]
```diff
@@ -21,6 +21,8 @@
     <exclude name="SystemPrintln"/>
     <exclude name="GuardLogStatement"/>
     <exclude name="AvoidPrintStackTrace"/>
+    <exclude name="UseVarargs"/>
```
Comment (axthosarouris) [inline config/pmd/ruleset.xml:24]: Rules violated by DTO classes that reflect json structures

File: src/main/java/no/unit/nva/doi/transformer/AbstractConverter.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package no.unit.nva.doi.transformer;
+
+import java.net.URI;
+import java.util.stream.Stream;
+import no.unit.nva.model.Organization;
+import no.unit.nva.model.PublicationDate;
+import no.unit.nva.model.PublicationStatus;
+
+public class AbstractConverter {
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/AbstractConverter.java:9]: Common methods between the DataciteConverter and the CrossrefConverter

File: src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java [hunk 3]
```diff
@@ -10,88 +17,61 @@
 import no.unit.nva.model.NameType;
 import no.unit.nva.model.Organization;
 import no.unit.nva.model.Publication;
-import no.unit.nva.model.PublicationDate;
-import no.unit.nva.model.PublicationStatus;
 import no.unit.nva.model.PublicationType;
 
-import java.net.URI;
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java:17]: Moved to the parent AbstractConverter

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java [hunk 4]
```diff
@@ -1,47 +1,51 @@
 package no.unit.nva.doi.transformer;
 
+import static org.apache.http.HttpHeaders.CONTENT_TYPE;
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/MainHandler.java:3]: Automatic rearrangement of imports

File: src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/Affiliation.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package no.unit.nva.doi.transformer.model.crossrefmodel;
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/Affiliation.java:1]: Crossref json model class.

File: src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/Author.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
+package no.unit.nva.doi.transformer.model.crossrefmodel;
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/Author.java:1]: Crossref model class

File: src/main/java/no/unit/nva/doi/transformer/Environment.java [hunk 7]
```diff
@@ -7,8 +7,8 @@
     /**
      * Get environment variable.
      *
-     * @param name  name of environment variable
-     * @return  optional with value of environment variable
+     * @param name name of environment variable
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/Environment.java:10]: Autoformating

File: src/main/java/no/unit/nva/doi/transformer/GatewayResponse.java [hunk 8]
```diff
@@ -16,15 +15,15 @@
     /**
      * Constructor for GatewayResponse.
      *
-     * @param body  body of response
-     * @param headers   http headers for response
-     * @param statusCode    status code for response
+     * @param body       body of response
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/GatewayResponse.java:18]: Autoformatting

File: src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/ContentDomain.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package no.unit.nva.doi.transformer.model.crossrefmodel;
+
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/ContentDomain.java:2]: Crossref model class

File: src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/CrossRefDocument.java [hunk 10]
```diff
@@ -0,0 +1,372 @@
+package no.unit.nva.doi.transformer.model.crossrefmodel;
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/CrossRefDocument.java:1]: Crossref model class

File: src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/CrossrefApiResponse.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
+package no.unit.nva.doi.transformer.model.crossrefmodel;
+
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/transformer/model/crossrefmodel/CrossrefApiResponse.java:2]: Nva-doi-lambda and nva-doi-fetch, they just add a location header and forward the response from Crossref as it is. So, we need to parse the original response from Crossref, which contains some other additional info

File: src/test/java/no/unit/nva/doi/transformer/EnvironmentTest.java [hunk 12]
```diff
@@ -1,19 +1,17 @@
 package no.unit.nva.doi.transformer;
 
+import static org.junit.Assert.assertEquals;
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/doi/transformer/EnvironmentTest.java:3]: Autoformating

File: src/test/java/no/unit/nva/doi/transformer/MainHandlerTest.java [hunk 13]
```diff
@@ -1,19 +1,24 @@
 package no.unit.nva.doi.transformer;
 
+import static java.util.Collections.singletonMap;
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/doi/transformer/MainHandlerTest.java:3]: Autoformating

File: src/test/java/no/unit/nva/doi/transformer/MainHandlerTest.java [hunk 14]
```diff
@@ -66,42 +76,39 @@ public void testDefaultConstructor() {
 
     @Test
     public void testOkResponse() throws IOException {
-        DataciteResponseConverter converter = new DataciteResponseConverter();
-        Context context = getMockContext();
-        MainHandler mainHandler = new MainHandler(objectMapper, converter, environment);
-        OutputStream output = new ByteArrayOutputStream();
+        MainHandlerWithAttachedOutputStream mainHandlerWithOutput = new MainHandlerWithAttachedOutputStream();
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/doi/transformer/MainHandlerTest.java:78]: Refactor repeated code

File: src/test/resources/datacite_response.xml [hunk 15]
```diff
@@ -1,28 +1,33 @@
...
-        <publicationDate dateType="Updated">2017-03-10T01:41:02Z</publicationDate>
-    </dates>
-    <relatedIdentifiers>
-        <relatedIdentifier relatedIdentifierType="ISSN" relationType="IsPartOf" resourceTypeGeneral="Collection">00189219</relatedIdentifier>
-    </relatedIdentifiers>
-    <version/>
-    <descriptions>
-        <description descriptionType="SeriesInformation">Proceedings of the IEEE, 87(7), 1208-1227</description>
-    </descriptions>
+<publication xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (axthosarouris) [inline src/test/resources/datacite_response.xml:2]: Autoformating

File: build.gradle [hunk 16]
```diff
@@ -27,12 +27,16 @@ dependencies {
     implementation group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.5.10'
     implementation group: 'org.zalando', name: 'problem', version: '0.23.0'
     implementation group: 'org.zalando', name: 'jackson-datatype-problem', version: '0.23.0'
-    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
+
+    //Used for different testing utilities such as reading from resources
+    testImplementation 'com.github.BIBSYSDEV:aws-build-tools:v1.5.5'
```
Comment (janbj) [inline build.gradle]: We should try to be consistent on format here.

File: src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+public class CrossRefConverter extends AbstractConverter {
+
+    public static final String NOT_A_JOURNAL_ARTICLE_ERROR = "The entry is not a journal article";
+    public static final String INVALID_ENTRY_ERROR = "The entry is empty or has no title";
+    protected static Map<String, Integer> ordinals;
+
+    public static String JOURNAL_ARTICLE = "journal-article";
+
+    static {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java]: Maybe add a comment about why we need this?

File: src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        } else {
+            throw new IllegalArgumentException(NOT_A_JOURNAL_ARTICLE_ERROR);
+        }
+    }
+
+    private String extractTitle(CrossRefDocument document) {
+        return getMainTitle(document.getTitle().stream());
+    }
+
+    private Optional<PublicationDate> extractDate(CrossRefDocument document) {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java:88]: Maybe a good idea to put in a reference to any external document with mapping rules or even write them here in coments.

File: src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+        Identity identity = new Identity.Builder()
+            .withName(toName(author.getFamilyName(), author.getGivenName()))
+            .build();
+        return new Contributor.Builder()
+            .withIdentity(identity)
+            .withSequence(parseSequence(author.getSequence()))
+            .build();
+    }
+
+    private int parseSequence(String sequence) {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java]: Add comment about what we want to achive here.

File: src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java [hunk 20]
```diff
@@ -10,88 +17,61 @@
...
+                .withName(toName(dataciteCreator))
+                .withNameType(NameType.lookup(dataciteCreator.getNameType()))
+                .build()
+            )
+            .withAffiliations(toAffilitation(dataciteCreator.getAffiliation()))
+            .withSequence(sequence)
+            .build();
     }
 
     protected List<Organization> toAffilitation(List<DataciteAffiliation> affiliation) {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java]: AffilitationS?

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java [hunk 21]
```diff
@@ -65,42 +69,90 @@ public MainHandler(ObjectMapper objectMapper, DataciteResponseConverter converte
...
-        DataciteResponse dataciteResponse;
         String owner;
         String orgNumber;
+        String body = null;
+        String contentLocation;
         try {
             JsonNode event = objectMapper.readTree(input);
-            String body = event.get(BODY).textValue();
-            dataciteResponse = objectMapper.readValue(body, DataciteResponse.class);
+            body = extractBody(event);
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/MainHandler.java]: extractRequestBody to make it more obvious

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java [hunk 22]
```diff
@@ -65,42 +69,90 @@ public MainHandler(ObjectMapper objectMapper, DataciteResponseConverter converte
...
-    private Map<String, String> headers() {
+    private Publication convertFromDatacite(String body, Instant now, String owner, UUID uuid, URI publisherId)
+        throws JsonProcessingException {
+        DataciteResponse dataciteResponse = objectMapper.readValue(body, DataciteResponse.class);
+        return converter.toPublication(dataciteResponse, now, uuid, owner, publisherId);
+    }
+
+    private Publication convertFromCrossRef(String body, Instant now, String owner, UUID identifier,
+                                            URI publisherId) throws JsonProcessingException {
+        CrossRefConverter converter = new CrossRefConverter();
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/MainHandler.java]: Create converter in constructor?

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java [hunk 23]
```diff
@@ -114,12 +166,12 @@ public void handleRequest(InputStream input, OutputStream output, Context contex
      */
     public static ObjectMapper createObjectMapper() {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/transformer/MainHandler.java:181]: Maybe we should move this into the nva-datamodel-java project before the 0.4 release?

**Review by janbj — DISMISSED** [review]
Looks good, just a few comments. :)

File: src/main/java/no/unit/nva/doi/transformer/AbstractConverter.java [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+
+    public static final PublicationStatus DEFAULT_NEW_PUBLICATION_STATUS = PublicationStatus.NEW;
+
+    protected String toName(String familyName, String givenName) {
+        return String.join(", ", familyName, givenName);
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/transformer/AbstractConverter.java]: Constant

File: src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java (hunk 18 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java:88]: The conversion is described in the class CrossRef date through documentation and tests

File: build.gradle [hunk 25]
```diff
@@ -27,12 +27,16 @@ dependencies {
...
     implementation group: 'org.zalando', name: 'jackson-datatype-problem', version: '0.23.0'
-    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
+
+    //Used for different testing utilities such as reading from resources
+    testImplementation group:'com.github.BIBSYSDEV', name:'aws-build-tools', version:'v1.5.5'
+
     testImplementation group: 'org.mockito', name: 'mockito-core', version: '3.2.4'
     testImplementation group: 'com.github.stefanbirkner', name: 'system-rules', version: '1.19.0'
-    testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
+    testImplementation "org.junit.jupiter:junit-jupiter:${junit5Version}"
```
Comment (brinxmat) [inline build.gradle]: Consistency

File: build.gradle [hunk 26]
```diff
@@ -27,12 +27,16 @@ dependencies {
...
     testCompileOnly 'junit:junit:4.12'
     testRuntimeOnly 'org.junit.vintage:junit-vintage-engine:5.1.0'
+    testImplementation 'org.hamcrest:hamcrest:2.2'
```
Comment (brinxmat) [inline build.gradle]: Consistency

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java [hunk 27]
```diff
@@ -1,106 +1,162 @@
...
-            dataciteResponse = objectMapper.readValue(body, DataciteResponse.class);
+            body = extractRequestBody(event);
+            contentLocation = extractContentLocationHeader(event);
             owner = getClaimValueFromRequestContext(event, CUSTOM_FEIDE_ID);
             orgNumber = getClaimValueFromRequestContext(event, CUSTOM_ORG_NUMBER);
         } catch (Exception e) {
             e.printStackTrace();
             objectMapper.writeValue(output, new GatewayResponse<>(objectMapper.writeValueAsString(
-                    Problem.valueOf(BAD_REQUEST, e.getMessage())), headers(), SC_BAD_REQUEST));
+                Problem.valueOf(BAD_REQUEST, e.getMessage())), responseHeaders(), SC_BAD_REQUEST));
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/transformer/MainHandler.java]: application/problem+json

**Review by brinxmat — DISMISSED** [review]

File: build.gradle (hunk 16 above)
Comment (axthosarouris) [reply build.gradle]: done.

File: src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java (hunk 17 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java]: done.

File: src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java (hunk 19 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java]: done

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java (hunk 21 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/MainHandler.java]: done

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java (hunk 23 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/MainHandler.java:181]: We could move it for the 0.5 release.

File: src/main/java/no/unit/nva/doi/transformer/AbstractConverter.java (hunk 24 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/AbstractConverter.java]: done

File: build.gradle (hunk 26 above)
Comment (axthosarouris) [reply build.gradle]: done

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java (hunk 27 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/MainHandler.java]: done

File: build.gradle [hunk 28]
```diff
@@ -27,12 +27,19 @@ dependencies {
...
+
+    //Used for different testing utilities such as reading from resources
+    testImplementation group:'com.github.BIBSYSDEV', name:'aws-build-tools', version:'v1.5.5'
+
     testImplementation group: 'org.mockito', name: 'mockito-core', version: '3.2.4'
     testImplementation group: 'com.github.stefanbirkner', name: 'system-rules', version: '1.19.0'
-    testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
-    testCompileOnly 'junit:junit:4.12'
-    testRuntimeOnly 'org.junit.vintage:junit-vintage-engine:5.1.0'
+    testImplementation group: 'org.junit.jupiter', name:'junit-jupiter', version:"${junit5Version}"
```
Comment (brinxmat) [inline build.gradle]: You can refer to the defined variable directly, without concatenation

**Review by brinxmat — DISMISSED** [review]

File: src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java (hunk 20 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java]: done

File: src/main/java/no/unit/nva/doi/transformer/MainHandler.java (hunk 22 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/transformer/MainHandler.java]: done

File: build.gradle (hunk 28 above)
Comment (axthosarouris) [reply build.gradle]: done

**Review by brinxmat — APPROVED** [review]

**Pull request merged**